### PR TITLE
Move cache to Resources Manager

### PR DESCRIPTION
**Pull request by frnmagalhaes**
Still have to use locks (maybe discuss this offline) on the ResourcesManager view cache

### CODE REVIEW

File: WebViewControl/ResourcesManager.cs [hunk 1]
```diff
@@ -9,6 +9,28 @@ namespace WebViewControl {
...
+            public string Extension;
+            public Stream ResourceStream;
+
+            public Resource(string extension, Stream resourceStream) {
+                Extension = extension;
+                ResourceStream = resourceStream;
+            }
+        }
+
+        public static Resource LoadEmbeddedResource(Uri url) {
```
Comment (joaompneves) [inline WebViewControl/ResourcesManager.cs]: 1) Move this method after the TryGetResourceWithFullPath
2) Rename it to TryGetResource
3) Remove the Resource class and create an out param for the extension
4) Create an overload internal method without the out param

File: WebViewControl/ResourcesManager.cs [hunk 2]
```diff
@@ -9,6 +9,28 @@ namespace WebViewControl {
 
     public static class ResourcesManager {
 
+        private static readonly AssemblyCache cache = new AssemblyCache();
+
+        public class Resource {
```
Comment (joaompneves) [inline WebViewControl/ResourcesManager.cs]: Drop this

File: WebViewControl/AssemblyCache.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace WebViewControl {
+
+    internal class AssemblyCache {
+
+        private Dictionary<string, Assembly> assemblies;
+        private bool newAssembliesLoaded = true;
+
```
Comment (joaompneves) [inline WebViewControl/AssemblyCache.cs]: create a lock object 
private readonly object SyncRoot { get; } = new object();

File: WebViewControl/AssemblyCache.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        internal Assembly ResolveResourceAssembly(Uri resourceUrl) {
+            if (assemblies == null) {
+                assemblies = new Dictionary<string, Assembly>();
```
Comment (joaompneves) [inline WebViewControl/AssemblyCache.cs]: if (assemblies == null) {
lock(SyncRoot) {
  if (assemblies == null) {
      assemblies = new Dictionary<string, Assembly>();
      ...

File: WebViewControl/AssemblyCache.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                AppDomain.CurrentDomain.AssemblyLoad += OnAssemblyLoaded;
+            }
+
+            var assemblyName = ResourceUrl.GetEmbeddedResourceAssemblyName(resourceUrl);
+            var assembly = GetAssemblyByName(assemblyName);
+
+            if (assembly == null) {
+                if (newAssembliesLoaded) {
```
Comment (joaompneves) [inline WebViewControl/AssemblyCache.cs]: if (newAssembliesLoaded) {
lock(SyncRoot) {
if (newAssembliesLoaded) {

File: WebViewControl/WebView.CefResourceRequestHandler.cs [hunk 6]
```diff
@@ -73,70 +69,18 @@ protected class CefResourceRequestHandler : ResourceRequestHandler, IDisposable
...
             }
 
-            private void OnAssemblyLoaded(object sender, AssemblyLoadEventArgs args) {
-                newAssembliesLoaded = true;
-            }
-
-            public void Dispose() {
-                AppDomain.CurrentDomain.AssemblyLoad -= OnAssemblyLoaded;
-            }
+            public void Dispose() { }
```
Comment (joaompneves) [inline WebViewControl/WebView.CefResourceRequestHandler.cs]: Remove

Comment (frnmagalhaes) [reply WebViewControl/WebView.CefResourceRequestHandler.cs]: are we ever going to unsubscribe the event?

Comment (frnmagalhaes) [reply WebViewControl/WebView.CefResourceRequestHandler.cs]: WebView.cs is expecting the RequestHandler to be IDisposable

File: WebViewControl/ResourcesManager.cs (hunk 1 above)
Comment (frnmagalhaes) [reply WebViewControl/ResourcesManager.cs]: we don't need the extension in the SS side though

File: WebViewControl/WebView.CefResourceRequestHandler.cs (hunk 6 above)
Comment (joaompneves) [reply WebViewControl/WebView.CefResourceRequestHandler.cs]: Does not need anymore

File: WebViewControl/AssemblyCache.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Reflection;
+
+namespace WebViewControl {
+
+    internal class AssemblyCache {
+
+        private static object SyncRoot { get; } = new object();
```
Comment (joaompneves) [inline WebViewControl/AssemblyCache.cs]: readonly ... not static

File: WebViewControl/ResourcesManager.cs [hunk 8]
```diff
@@ -74,6 +76,20 @@ public static class ResourcesManager {
             return InternalTryGetResource(assemblyName, resourcePath.First(), resourcePath.Skip(1), false);
         }
 
+        public static Stream TryGetResource(Uri url, out string extension) {
```
Comment (joaompneves) [inline WebViewControl/ResourcesManager.cs]: this one should be internal, the other public

File: WebViewControl/WebView.CefResourceRequestHandler.cs (hunk 6 above)
Comment (joaompneves) [reply WebViewControl/WebView.CefResourceRequestHandler.cs]: Doesn't neeed to

File: WebViewControl/AssemblyCache.cs (hunk 7 above)
Comment (frnmagalhaes) [reply WebViewControl/AssemblyCache.cs]: the modifier readonly is not valid for this item. I've simply removed static. In WebView.cs it's just private object

**Pull request merged**